pkg/operator/k8sutil/node.go

```diff
@@ -364,3 +364,21 @@ func GenerateNodeAffinity(nodeAffinity string) (*v1.NodeAffinity, error) {
 	}
 	return newNodeAffinity, nil
 }
+
+func GetKubernetesReadyNodes(clientset kubernetes.Interface) ([]v1.Node, error) {
+	nodes, err := clientset.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var readyNodes []v1.Node
+	for _, node := range nodes.Items {
+		for _, nodeCondition := range node.Status.Conditions {
+			if nodeCondition.Type == v1.NodeReady && nodeCondition.Status == "True" {
```

Comment:
If someone sets the [node affinity](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/cluster.yaml#L80-L87) for the mons, this basic test for k8s nodes won't be sufficient. Did you try the `k8sutil.ValidNode()` method? Then you can take the placement for the mon into account.

Reply:
@travisn is this PR merely recreating the placement logic that Noah worked so hard to get rid of? The point of the canary pods is to let k8s tell Rook whether mons can be scheduled instead of implementing a bunch of custom logic in Rook that duplicates what k8s does. Some basic sanity checks seem okay to me, but I would advise against the tempation to try to "optimize" too much here.

Reply:
Great question, this has been troubling me as well... The whole purpose of ripping out the scheduling logic was to let K8s take care of it and now the complexity is returning.
@jmolmo How about if we think about this a bit differently? The problem is really that the canary pod stays pending indefinitely and never goes away. How about if the operator still always started the canary pod, but then aborted the failover and deleted the canary pod if it stays in pending state for too long (~5 minutes)? This way we don't introduce any scheduling code, but we clean up gracefully.

Reply:
That sounds much better to me. I didn't realize we weren't cleaning up canaries if they were unschedulable.

Reply:
@BlaineEXE, @travisn : 

Well, really what we have had all the time is more or less the same that @travisn  has described as possible solution.
@noahdesu  did an excellent work here and this situation is also covered. The canary-pod is deleted regularly, and a new canary pod is created as consequence of the restart of the failover procedure. This cycle is infinite, and for that reason it seems that a canary pod is in pending state forever( it is created, deleted in each failover cycle).
I think that this behavior is correct, and in any case the only thing what we can do is to reduce the time checking if the canary pod is ok or not (reducing canaryRetries var). 

The solution i have implemented in this PR is to avoid the failover procedure when we know in advance that there are no nodes available to complete this process successfully. And therefore, the pending mon canary pod (different instances...) will be always present.

What should i do? 
- Should i continue with this solution improving the implementation taking into account also the mon placement? 

- It does not matter to have the pending canary pod being recreated all the time? should we close #3764  as  "won't fix"?
Note: If there are Nodes available, the mon canary pod is always deleted, or at least i cannot be able to reproduce the situation where the mon canary pod is not deleted.

**Details about the failover cycle with an example:**

Once the timeout period for acquire quorum pass the failover for the mon starts:
```

2019-12-20 08:46:51.450088 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
2019-12-20 08:46:51.450113 I | op-mon: Failing over monitor "c"
2019-12-20 08:46:51.450138 I | op-mon: starting new mon: &{ResourceName:rook-ceph-mon-e DaemonName:e PublicIP: Port:6789 DataPathMap:0xc000ff0750}
2019-12-20 08:46:51.454986 I | op-mon: sched-mon: created canary deployment rook-ceph-mon-e-canary
2019-12-20 08:46:51.474624 I | op-mon: sched-mon: waiting for canary pod creation rook-ceph-mon-e-canary
```

Here we are in the situation where the canary pod is created, but it cannot be scheduled because there are no nodes available. 

We are 15 minutes waiting because we are retrying the mon-canary pod creation for 
canaryRetries(180) * canaryRetryDelaySeconds(5) = 900 seconds = 15 minutes

Once we have exhausted the number of retries, the canary pod is cleaned:

```
2019-12-20 09:01:53.512780 I | op-mon: assignmon: cleaning up canary deployment rook-ceph-mon-e-canary and canary pvc 
2019-12-20 09:01:53.512832 I | op-k8sutil: removing deployment rook-ceph-mon-e-canary if it exists
2019-12-20 09:01:53.520543 I | op-k8sutil: Removed deployment rook-ceph-mon-e-canary
2019-12-20 09:01:53.526424 I | op-k8sutil: rook-ceph-mon-e-canary still found. waiting...
2019-12-20 09:01:55.531978 I | op-k8sutil: confirmed rook-ceph-mon-e-canary does not exist

```
And a message reports the situation:

`2019-12-20 09:01:55.532043 E | op-mon: failed to failover mon "c". failed to place new mon on a node: assignmon: error scheduling monitor: sched-mon: canary pod scheduling failed retries
`

And a new failover cycle starts when the mon health check is launched again (each <HealthCheckInterval> = 45 seconds)

`
2019-12-20 09:02:41.001187 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
`

Reply:
@jmolmo Agreed that the operator is doing the right thing by trying to re-create the canary pod with each failover attempt. The scenario that could still use a bug fix is [this one](https://rook-io.slack.com/archives/CK9CF5H2R/p1576862700036500). If we're not anymore in the failover code path, the canary pod won't be removed. So the fix could be that if there is no failover needed and all mons are running, the health check can remove any extra canary pods since they aren't needed anymore.

Reply:
@travisn: probably wong link: ( [this one](https://rook-io.slack.com/archives/CK9CF5H2R/p1576862700036500) ) ... encrypted OSDs?